### DISCUSSION TITLE
Please give me a simple example of a false negative of casadi.is_linear

### DISCUSSION THREAD

**Discussion by eike-fokken** [casadi/casadi#3578] [Q&A] [answered]
In the thread

https://github.com/casadi/casadi/discussions/3382

I gave an implementation of a function checking linearity of a `casadi.Function`. It sometimes (always?) correctly identifies a linear function that is a false negative of `casadi.is_linear`.

Now I am testing it and the false negative function I have is huge and constructed in a complicated way.

Could someone give me a hopefully small linear function that is incorrectly identified as non-linear by `casadi.is_linear`?

**Comment by mschloeder-abb** [comment, marked as answer]
Check out this example. Here, is_linear returns False when the evaluated function is used instead of the expression behind the function, which is in fact linear in x:

a = casadi.MX.sym("a", 1)
b = casadi.MX.sym("b", 1)
test = casadi.Function(
    "test", [a, b], [a*b]
)

x = casadi.MX.sym("x", 1)
p = casadi.MX.sym("p", 1)

casadi.is_linear(test(x, p), x)

**Reply by eike-fokken** [reply]
That works. To me it looks like casadi cannot discern linearity of an expression, if that expression is created by evaluating a `casadi.Function` on some casadi variables. @jgillis : Could you comment on this and do you see a way to make it possible to peek into functions to check for linearity?

Also: I think, also the solution I gave in the other thread thinks that this function is non-linear, so nothing gained there.

**Reply by jgillis** [reply]
This is reported in #2030 - the way forward is to have an expression transformation that inlines Function calls for the purpose of a linearity check. You can try this on your model by constructing all Functions with force_inline True option:

```python
a = casadi.MX.sym("a", 1)
b = casadi.MX.sym("b", 1)
test = casadi.Function(
"test", [a, b], [a*b],
{"always_inline":True}
)

x = casadi.MX.sym("x", 1)
p = casadi.MX.sym("p", 1)

casadi.is_linear(test(x, p), x)
```
